src/MSBuildBinLogQuery/Construction/GraphBuilder.cs

```diff
@@ -65,10 +68,75 @@ private void TargetStarted(object sender, TargetStartedEventArgs args)
             }
         }
 
-        private Project GetParentNode(ProjectStartedEventArgs args)
+        private void TaskStarted(object sender, TaskStartedEventArgs args)
+        {
+            var project = Build.Projects[args.BuildEventContext.ProjectInstanceId];
+            var target = project.TargetsById[args.BuildEventContext.TargetId];
+
+            target.AddOrGetTask(args.BuildEventContext.TaskId, args.TaskName, args.TaskFile);
+        }
+
+        private void MessageRaised(object sender, BuildMessageEventArgs args)
+        {
+            var containingComponent = GetContainingComponent(
+                args.BuildEventContext.ProjectInstanceId,
+                args.BuildEventContext.TargetId,
+                args.BuildEventContext.TaskId);
+            var message = new Message(args.Message, containingComponent, args.Importance);
+
+            containingComponent.AddMessage(message);
+        }
+
+        private void WarningRaised(object sender, BuildWarningEventArgs args)
+        {
+            var containingComponent = GetContainingComponent(
+                args.BuildEventContext.ProjectInstanceId,
+                args.BuildEventContext.TargetId,
+                args.BuildEventContext.TaskId);
+            var warning = new Warning(args.Message, containingComponent);
+
+            containingComponent.AddWarning(warning);
+        }
+
+        private void ErrorRaised(object sender, BuildErrorEventArgs args)
+        {
+            var containingComponent = GetContainingComponent(
+                args.BuildEventContext.ProjectInstanceId,
+                args.BuildEventContext.TargetId,
+                args.BuildEventContext.TaskId);
+            var error = new Error(args.Message, containingComponent);
+
+            containingComponent.AddError(error);
+        }
+
+        private Project GetParentProject(ProjectStartedEventArgs args)
         {
             var parentId = args.ParentProjectBuildEventContext.ProjectInstanceId;
-            return Projects.TryGetValue(parentId, out var parent) ? parent : null;
+            return Build.Projects.TryGetValue(parentId, out var parent) ? parent : null;
+        }
+
+        private BuildComponent GetContainingComponent(int projectInstanceId, int targetId, int taskId)
```

Comment:
Consider simplifying this by taking a `BuildEventContext` instead of the three arguments? Or are there cases where you don't have a whole one?